ArmPkg/Drivers/SmmuDxe/IoMmu.c

```diff
@@ -0,0 +1,523 @@
+  SMMUV3_BUILD_CMD_TLBI_EL2_ALL (&Command);
+  SmmuV3SendCommand (mSmmu, &Command);
+  // Issue a CMD_SYNC command to guarantee that any previously issued TLB
+  // invalidations (CMD_TLBI_*) are completed (SMMUv3.2 spec section 4.6.3).
+  SMMUV3_BUILD_CMD_SYNC_NO_INTERRUPT (&Command);
+  SmmuV3SendCommand (mSmmu, &Command);
+
+  if (MapInfo != NULL) {
+    FreePool (MapInfo);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Free a buffer allocated by IoMmuAllocateBuffer.
+
+  @param [in]  This          Pointer to the IOMMU protocol instance.
+  @param [in]  Pages         The number of pages to free.
+  @param [in]  HostAddress   The host address to free.
+
+  @retval EFI_SUCCESS        Success.
+**/
+EFI_STATUS
+EFIAPI
+IoMmuFreeBuffer (
+  IN  EDKII_IOMMU_PROTOCOL  *This,
+  IN  UINTN                 Pages,
+  IN  VOID                  *HostAddress
+  )
+{
+  return gBS->FreePages ((EFI_PHYSICAL_ADDRESS)(UINTN)HostAddress, Pages);
+}
+
+/**
+  Allocate a buffer for use with the IOMMU.
+
+  @param [in]      This          Pointer to the IOMMU protocol instance.
+  @param [in]      Type          The type of allocation to perform.
+  @param [in]      MemoryType    The type of memory to allocate.
+  @param [in]      Pages         The number of pages to allocate.
+  @param [in, out] HostAddress   On input, the desired host address. On output, the allocated host address.
+  @param [in]      Attributes    The memory attributes to use for the allocation.
+
+  @retval EFI_SUCCESS            Success.
+  @retval EFI_INVALID_PARAMETER  Invalid parameter.
+  @retval EFI_OUT_OF_RESOURCES   Out of resources.
+**/
+EFI_STATUS
+EFIAPI
+IoMmuAllocateBuffer (
+  IN     EDKII_IOMMU_PROTOCOL  *This,
+  IN     EFI_ALLOCATE_TYPE     Type,
+  IN     EFI_MEMORY_TYPE       MemoryType,
+  IN     UINTN                 Pages,
+  IN OUT VOID                  **HostAddress,
+  IN     UINT64                Attributes
+  )
+{
+  EFI_STATUS            Status;
+  EFI_PHYSICAL_ADDRESS  PhysicalAddress;
+
+  Status = gBS->AllocatePages (
+                  Type,
+                  MemoryType,
+                  Pages,
+                  &PhysicalAddress
+                  );
+  if (!EFI_ERROR (Status)) {
+    *HostAddress = (VOID *)(UINTN)PhysicalAddress;
+  }
+
+  return Status;
+}
+
+/**
+  Set the R/W attributes for a device handle.
+
+  @param [in]  This          Pointer to the IOMMU protocol instance.
+  @param [in]  DeviceHandle  The device handle to set attributes for.
+  @param [in]  Mapping       The mapping to set attributes for.
+  @param [in]  IoMmuAccess   The IOMMU access attributes for R/W.
+
+  @retval EFI_SUCCESS        Success.
+**/
+EFI_STATUS
+EFIAPI
+IoMmuSetAttribute (
+  IN EDKII_IOMMU_PROTOCOL  *This,
+  IN EFI_HANDLE            DeviceHandle,
+  IN VOID                  *Mapping,
+  IN UINT64                IoMmuAccess
+  )
+{
+  EFI_STATUS      Status;
+  IOMMU_MAP_INFO  *MapInfo;
+
+  if (Mapping == NULL) {
+    return EFI_SUCCESS;
+  }
+

```

Comment:
Should you also validate the `IoMmuAccess` has valid bits before passing it on? Probably better to fail it with `EFI_INVALID_PARAMETER` here?

Reply:
done